components/merkledb/src/proof_map_index/proof.rs

```diff
@@ -551,20 +561,23 @@ where
     /// let proof = map.get_proof(h2);
     /// let checked_proof = proof.check().unwrap();
     /// assert_eq!(checked_proof.entries().collect::<Vec<_>>(), vec![(&h2, &200u32)]);
-    /// assert_eq!(checked_proof.root_hash(), map.object_hash());
+    /// assert_eq!(checked_proof.object_hash(), map.object_hash());
     /// ```
     ///
     /// [`ProofMapIndex`]: struct.ProofMapIndex.html
-    pub fn check(self) -> Result<CheckedMapProof<K, V>, MapProofError> {
+    pub fn check(&self) -> Result<CheckedMapProof<K, V>, MapProofError> {
         self.precheck()?;
-        let (mut proof, entries) = (self.proof, self.entries);
 
-        proof.extend(entries.iter().filter_map(|e| {
-            e.as_kv().map(|(k, v)| MapProofEntry {
-                path: ProofPath::new(k),
-                hash: HashTag::hash_leaf(&v.to_bytes()),
+        let mut proof: Vec<_> = self.proof.iter().map(Cow::Borrowed).collect();
```

Comment:
I've benchmarked variants with `Cow` / with cloning `self.proof` locally a little, but didn't get conclusive results. The performance relatively equal in both cases.

As a side note, verifying `ListProof`s seems to be ~5x faster than verifying `MapProof`s with the same number of entries in the list / map. I'm not sure why; there is greater amount of data hashed in `MapProof`s, but the overhead shouldn't be *that* large.